Create new GridList component based on old GridView

ethanshar opened the pull request:
## Description
Create new GridList component based on old GridView

## Changelog
Create new GridList component based on old GridView features

M-i-k-e-l commented on src/components/gridList/index.tsx:
```diff
@@ -0,0 +1,86 @@
+
+  const _renderItem = useCallback((...args) => {
+    // @ts-expect-error
+    return <View style={[itemContainerStyle]}>{renderItem?.(...args)}</View>;
```
No need for the `[]`

ethanshar replied:
True, removed

M-i-k-e-l commented on src/components/gridList/index.tsx:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * List padding (used for item size calculation)
+   */
+  listPadding?: number;
```
The `listPadding` is not acting as expected, for example change the screen to use:
```
numColumns={2}
// maxItemWidth={140}
itemSpacing={Spacings.s10}
listPadding={Spacings.s2}
```
There is no padding on the right side.

Or:
```
numColumns={2}
// maxItemWidth={140}
itemSpacing={Spacings.s10}
// listPadding={Spacings.s2}
```
It looks like the right item is a little out of the screen

ethanshar replied:
Well, technically the `listPadding` prop goes together with setting the list `contentContainerStyle` with padding. 
But I see your point, I think I'll inject the padding in the list myself so the user won't have to worry about it

M-i-k-e-l commented on demo/src/screens/componentScreens/GridListScreen.tsx:
```diff
@@ -0,0 +1,60 @@
+          <Text>{item.name}</Text>
+          <Text>{item.formattedPrice}</Text>
+          {item.inventory.status === 'Out of Stock' && (
+            <Text text90M red30>
```
Consider using `$textDangerLight`

ethanshar replied:
Fixed

ethanshar commented on src/index.js:
```diff
@@ -69,6 +69,9 @@ export default {
+  get GridList() {
+    return require('./components/gridList').default;
+  },
   get GridView() {
```
Yes, definitely, I'll create a separate ticket for that

ethanshar commented on src/components/gridList/index.tsx:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * whether to keep the items initial size when orientation changes,
+   * in which case the apt number of columns will be calculated automatically.
+   * Ignored when passing 'maxItemWidth'
```
Left overs from the old grid view implementation, removed

ethanshar commented on src/components/gridList/index.tsx:
```diff
@@ -0,0 +1,86 @@
+  const containerWidth = useMemo(() => {
+    return (props.containerWidth ?? Constants.screenWidth) - listPadding * 2;
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [listPadding, orientation]);
```
Added containerWidth (the comment still relevant because of the orientation dep) 
Anyway, the code has been moved to useGridLayout hook

ethanshar commented:
@M-i-k-e-l I fixed/comment everything, notice that I did a small refactor and moved the logic and calculation of the grid to useGridLayout hook (this is relevant to the next PR of the SortableGridList)

lidord-wix commented on src/components/gridList/useGridLayout.ts:
```diff
@@ -0,0 +1,49 @@
+
+  const numberOfColumns = useMemo(() => {
+    if (maxItemWidth) {
+      return Math.ceil((_containerWidth + itemSpacing) / (maxItemWidth + itemSpacing));
```
Is our interest here is (1) to maximize the number of columns or (2) to maximize the width of each item (and keep it less than the `maxItemWidth`)? 
if option 2 is the answer (and as far as I understand it is), we should use `Math.floor `here and not `Math.ceil`

ethanshar replied:
The idea is to max the item width to the maxItemWidth passed by the user. 
When I change to Math.floor I see that items width exceed the maxItemWidth passed in the example screen (I pass 140 and the items width become 169)

lidord-wix commented on src/components/gridList/useGridLayout.ts:
```diff
@@ -0,0 +1,49 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [numColumns, maxItemWidth, itemSpacing, keepItemSize ? _containerWidth : undefined]);
+
+  const itemSize = useMemo(() => {
```
I think it's better to call it 'itemWidth' since it does nothing with height.

ethanshar replied:
renamed

lidord-wix commented on src/components/gridList/useGridLayout.ts:
```diff
@@ -0,0 +1,49 @@
+    return {width: itemSize /* + itemSpacing */, marginRight: itemSpacing, marginBottom: itemSpacing};
+  }, [itemSize, itemSpacing]);
+
+  return {itemContainerStyle, numberOfColumns, itemSize};
```
I see that you return the `itemSize` but I don't see that we using it somewhere outside.

ethanshar replied:
Removed

lidord-wix commented on demo/src/screens/componentScreens/GridListScreen.tsx:
```diff
@@ -0,0 +1,60 @@
+
+  renderItem: GridListProps<typeof products[0]>['renderItem'] = ({item}) => {
+    return (
+      <Card flex>
```
Consider adding `$backgroundElevated` for the card background and `$textDefault` for the texts

lidord-wix reviewed (commented):
Looks good! 
I left few small comments :)

ethanshar commented on src/components/gridList/index.tsx:
```diff
@@ -0,0 +1,49 @@
+    maxItemWidth,
+    listPadding,
+    keepItemSize
+  });
```
passed 👌
